routing/cross_mwm_index_graph.hpp

```diff
@@ -1,138 +1,67 @@
   void GetEdgeList(Segment const & s, bool isOutgoing, std::vector<SegmentEdge> & edges);
-
-  void Clear();
-
+  void Clear() { m_connectors.clear(); }
+  TransitionPoints GetTransitionPoints(Segment const & s, bool isOutgoing);
+  bool HasCache(NumMwmId numMwmId) const { return m_connectors.count(numMwmId) != 0; }
 private:
-  struct TransitionSegments
-  {
-    std::map<Segment, std::vector<ms::LatLon>> m_ingoing;
-    std::map<Segment, std::vector<ms::LatLon>> m_outgoing;
-  };
-
-  void ResetCrossMwmGraph();
+  CrossMwmConnector const & GetCrossMwmConnectorWithTransitions(NumMwmId numMwmId);
+  CrossMwmConnector const & GetCrossMwmConnectorWithWeights(NumMwmId numMwmId);
 
-  /// \brief Inserts all ingoing and outgoing transition segments of mwm with |numMwmId|
-  /// to |m_transitionCache|.
-  void InsertWholeMwmTransitionSegments(NumMwmId numMwmId);
-
-  /// \brief Fills |borderCrosses| of mwm with |mapping| according to |s|.
-  /// \param mapping if |isOutgoing| == true |mapping| is mapping ingoing (from) border cross.
-  /// If |isOutgoing| == false |mapping| is mapping outgoing (to) border cross.
-  /// \note |s| and |isOutgoing| params have the same restrictions which described in
-  /// GetEdgeList() method.
-  void GetBorderCross(TRoutingMappingPtr const & mapping, Segment const & s, bool isOutgoing,
-                      std::vector<BorderCross> & borderCrosses);
-
-  TransitionSegments const & GetSegmentMaps(NumMwmId numMwmId);
-
-  std::vector<ms::LatLon> const & GetIngoingTransitionPoints(Segment const & s);
-  std::vector<ms::LatLon> const & GetOutgoingTransitionPoints(Segment const & s);
-
-  template <class Fn>
-  bool LoadWith(NumMwmId numMwmId, Fn && fn)
+  template <typename Fn>
+  CrossMwmConnector const & Deserialize(NumMwmId numMwmId, Fn && fn)
   {
-    platform::CountryFile const & countryFile = m_numMwmIds->GetFile(numMwmId);
-    TRoutingMappingPtr mapping = m_indexManager.GetMappingByName(countryFile.GetName());
-    CHECK(mapping, ("No routing mapping file for countryFile:", countryFile));
-
-    if (!mapping->IsValid())
-      return false; // mwm was not loaded.
-
-    auto const it = m_mappingGuards.find(numMwmId);
-    if (it == m_mappingGuards.cend())
-    {
-      m_mappingGuards[numMwmId] = make_unique<MappingGuard>(mapping);
-      mapping->LoadCrossContext();
-    }
-
-    fn(mapping);
-    return true;
+    MwmSet::MwmHandle handle = m_index.GetMwmHandleByCountryFile(m_numMwmIds->GetFile(numMwmId));
+    CHECK(handle.IsAlive(), ());
+    MwmValue * value = handle.GetValue<MwmValue>();
+    CHECK(value != nullptr, ("Country file:", m_numMwmIds->GetFile(numMwmId)));
+
+    FilesContainerR::TReader const reader =
+        FilesContainerR::TReader(value->m_cont.GetReader(CROSS_MWM_FILE_TAG));
+    ReaderSource<FilesContainerR::TReader> src(reader);
+    auto const p = m_connectors.emplace(numMwmId, CrossMwmConnector(numMwmId));
```

Comment:
Stop. It's impossible. But I see. This code either contains or mistake or inefficient. I'll fix it tomorrow morning carefully. Thanks.